libs/platform/ebpf_crypto_hash.c

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+#include "ebpf_platform.h"
+
+#include <bcrypt.h>
+
+#pragma comment(lib, "bcrypt")
+
+// This file contains the wrapper around the crypto API.
+
+typedef struct _ebpf_cryptographic_hash
+{
+    BCRYPT_ALG_HANDLE algorithm_handle;
+    BCRYPT_HASH_HANDLE hash_handle;
+} ebpf_cryptographic_hash_t;
+
+_Must_inspect_result_ ebpf_result_t
+ebpf_cryptographic_hash_create(const wchar_t* algorithm, _Outptr_ ebpf_cryptographic_hash_t** hash)
+{
+    ebpf_result_t result;
+    NTSTATUS nt_status;
+    ebpf_cryptographic_hash_t* local_hash =
+        (ebpf_cryptographic_hash_t*)ebpf_allocate(sizeof(ebpf_cryptographic_hash_t));
+    if (local_hash == NULL) {
+        result = EBPF_NO_MEMORY;
+        goto Done;
+    }
+
+    memset(local_hash, 0, sizeof(ebpf_cryptographic_hash_t));
+
+    nt_status = BCryptOpenAlgorithmProvider(&local_hash->algorithm_handle, algorithm, NULL, 0);
+    if (!NT_SUCCESS(nt_status)) {
+        result = EBPF_INVALID_ARGUMENT;
+        goto Done;
+    }
+
+    nt_status = BCryptCreateHash(local_hash->algorithm_handle, &local_hash->hash_handle, NULL, 0, NULL, 0, 0);
+    if (!NT_SUCCESS(nt_status)) {
+        result = EBPF_INVALID_ARGUMENT;
+        goto Done;
+    }
+
+    result = EBPF_SUCCESS;
+    *hash = local_hash;
+    local_hash = NULL;
+Done:
+    ebpf_cryptographic_hash_destroy(local_hash);
+    local_hash = NULL;
+    return result;
+}
+
+void
+ebpf_cryptographic_hash_destroy(_In_opt_ _Frees_ptr_opt_ ebpf_cryptographic_hash_t* hash)
```

Comment:
If you use `_Frees_ptr_opt_`, do you still need `_In_opt_` or is it redundant?  I think other places don't have both,